core/server/src/test/java/alluxio/worker/netty/DataServerBlockWriteHandlerTest.java

```diff
@@ -0,0 +1,214 @@
+    Assert.assertTrue(writeResponse instanceof RPCProtoMessage);
+    checkWriteResponse(writeResponse, Protocol.Status.Code.INTERNAL);
+    waitForChannelClose(channel);
+    Assert.assertTrue(!channel.isOpen());
+  }
+
+  /**
+   * Builds the write request.
+   *
+   * @param offset the offset
+   * @param len the length of the block
+   * @return the write request
+   */
+  private RPCProtoMessage buildWriteRequest(long offset, int len) {
+    Protocol.WriteRequest writeRequest =
+        Protocol.WriteRequest.newBuilder().setId(mBlockId).setOffset(offset).setSessionId(1L)
+            .setType(Protocol.RequestType.ALLUXIO_BLOCK).build();
+    DataBuffer buffer = null;
+    if (len > 0) {
+      ByteBuf buf = PooledByteBufAllocator.DEFAULT.buffer(len);
+      for (int i = 0; i < len; i++) {
+        byte value = (byte) (mRandom.nextInt() % Byte.MAX_VALUE);
+        buf.writeByte(value);
+        mChecksum += BufferUtils.byteToInt(value);
+      }
+      buffer = new DataNettyBufferV2(buf);
+    }
+    return new RPCProtoMessage(writeRequest, buffer);
+  }
+
+  /**
+   * Checks the given write response is expected and matches the given error code.
+   *
+   * @param writeResponse the write response
+   * @param codeExpected the expected error code
+   */
+  private void checkWriteResponse(Object writeResponse, Protocol.Status.Code codeExpected) {
+    Assert.assertTrue(writeResponse instanceof RPCProtoMessage);
+
+    Object response = ((RPCProtoMessage) writeResponse).getMessage();
+    Assert.assertTrue(response instanceof Protocol.Response);
+    Assert.assertEquals(codeExpected, ((Protocol.Response) response).getStatus().getCode());
+  }
+
+  /**
+   * Checks the file content matches expectation (file length and file checksum).
+   *
+   * @param size the file size in bytes
+   * @throws IOException if it fails to check the file content
+   */
+  private void checkFileContent(long size) throws IOException {
+    RandomAccessFile file = new RandomAccessFile(mFile, "r");
+    long checksumActual = 0;
+    long sizeActual = 0;
+    try {
+      while (true) {
+        checksumActual += BufferUtils.byteToInt(file.readByte());
+        sizeActual++;
+      }
+    } catch (EOFException e) {
+      // expected.
+    }
+    Assert.assertEquals(mChecksum, checksumActual);
+    Assert.assertEquals(size, sizeActual);
+  }
+
+  /**
+   * Waits for a response.
+   *
+   * @param channel the channel
+   * @return the response
+   */
+  private Object waitForResponse(EmbeddedChannel channel) {
+    Object writeResponse = null;
+    int timeRemaining = Constants.MINUTE_MS;
+    while (writeResponse == null && timeRemaining > 0) {
+      writeResponse = channel.readOutbound();
+      CommonUtils.sleepMs(10);
+      timeRemaining -= 10;
+    }
+    return writeResponse;
+  }
+
+  /**
+   * Waits for the channel to be closed.
+   *
+   * @param channel the channel
+   */
+  private void waitForChannelClose(EmbeddedChannel channel) {
```

Comment:
This is like the other one. This functionality should be pulled out to one place.